File: backend/api/authorization.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from ..services.autorization import (
    create_role_with_check, get_role, list_all_roles, assign_role_to_user_with_check, get_roles_for_user, user_has_role, remove_role_from_user
)
from ..schemas.authorization import RoleCreate, UserRoleCreate
from ..database import get_db
from typing import Optional

router = APIRouter()

@router.post("/role/create")
def create_role(role: RoleCreate, db: Session = Depends(get_db)):
    return create_role_with_check(db, role)

@router.get("/role/get")
def get_role_by_name(name: str, db: Session = Depends(get_db)):
    return get_role(db, name)

@router.get("/role/list")
def list_roles(db: Session = Depends(get_db)):
    return list_all_roles(db)

@router.post("/user-role/assign")
def assign_role(user_role: UserRoleCreate, db: Session = Depends(get_db)):
    return assign_role_to_user_with_check(db, user_role)

@router.get("/user-role/list/{user_id}")
def get_user_roles_api(user_id: int, db: Session = Depends(get_db)):
    return get_roles_for_user(db, user_id)

@router.get("/user-role/has-role")
def has_role(user_id: int, role_name: str, db: Session = Depends(get_db)):
    return {"has_role": user_has_role(db, user_id, role_name)}

@router.delete("/user-role/remove")
def remove_role(user_id: int, role_id: int, db: Session = Depends(get_db)):
    return {"removed": remove_role_from_user(db, user_id, role_id)}

File: frontend/src/api/authorization.ts
// API utility for authorization endpoints matching backend/api/authorization.py

export async function createRole(role: any) {
  const response = await fetch('/api/authorization/role/create', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(role),
  });
  if (!response.ok) throw new Error('Failed to create role');
  return response.json();
}

export async function getRoleByName(name: string) {
  const response = await fetch(`/api/authorization/role/get?name=${encodeURIComponent(name)}`);
  if (!response.ok) throw new Error('Failed to get role');
  return response.json();
}

export async function listRoles() {
  const response = await fetch('/api/authorization/role/list');
  if (!response.ok) throw new Error('Failed to list roles');
  return response.json();
}

export async function assignRoleToUser(userRole: any) {
  const response = await fetch('/api/authorization/user-role/assign', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(userRole),
  });
  if (!response.ok) throw new Error('Failed to assign role to user');
  return response.json();
}

export async function getRolesForUser(userId: number) {
  const response = await fetch(`/api/authorization/user-role/list/${userId}`);
  if (!response.ok) throw new Error('Failed to get roles for user');
  return response.json();
}

export async function userHasRole(userId: number, roleName: string) {
  const response = await fetch(`/api/authorization/user-role/has-role?user_id=${userId}&role_name=${encodeURIComponent(roleName)}`);
  if (!response.ok) throw new Error('Failed to check user role');
  return response.json();
}

export async function removeRoleFromUser(userId: number, roleId: number) {
  const response = await fetch(`/api/authorization/user-role/remove?user_id=${userId}&role_id=${roleId}`, {
    method: 'DELETE',
  });
  if (!response.ok) throw new Error('Failed to remove role from user');
  return response.json();
}
